### PR TITLE
Overflow and batched queue unification

### PR DESCRIPTION
**Pull request by vasilmkd**
Opening to test in the CI. I still need to adjust the comments in the runtime for the new queue and simplified worker thread state machine.

### CODE REVIEW

**Comment by djspiewak** [comment]
For anyone following the breadcrumbs on this PR and looking for a deeper explanation, this is a subtler change than it may seem from the outside. Prior to this PR, we had the following set of work queues:

- Local queues (on each worker thread, capped at 256 fibers)
- External queue (global to the pool, unbounded and shared)
- Batched queue (global to the pool, unbounded and shared, consisting of fiber batches of size 32 each)

New *internal* work (e.g. `IO#start` and `IO.cede`) is submitted to the local queues, while new *external* work (e.g. `IO.async` callback fulfillment) is submitted to the external queue. The batched queue was used when the local queue for a particular worker exceeded its limit of 256 (which can happen for several different reasons). Whenever the local queue overflows, 128 fibers are removed from that queue and bundled into four batches of size 32 each, then all of these are placed on the batched queue.

Every 64 `cede`s, each worker will poll *either* the external queue or the batched queue (alternating between the two) to see if there is any global work which must be attended. Thus, workers are heavily optimized for local work, and anything that ends up on the shared queues is relatively slow. The shared queues are split between externally- and internally-sourced work sources, where the fairness ratio is 32:1 in favor of the locally-sourced work (corresponding to the batch size), since we alternate polling the queues at an exactly 1:1 ratio.

This works well, but unfortunately it's a bit rigid. Applications which have a *lot* of externally-sourced work relative to their internally-sourced work may find that the 32:1 ratio is inappropriate and starves the external sourcing. Conversely, applications which have very little externally-sourced work may find that the internal work is not given due priority. This is also not a ratio which necessarily remains static, since applications may experience *bursts* of external work followed by long stretches of internal work. A more dynamic approach is more appropriate.

This PR effectively changes the ratio to a dynamic one by pushing everything onto a unified queue. We retain the batching property of the overflow spill, which improves contention considerably and puts a thumb on the scale toward internal sourcing, but by putting everything onto a single queue we allow for consecutive takes from externally-sourced tasks, as well as consecutive takes from internally-sourced tasks. If either ends up producing more tasks than the other, it will be more represented on the external queue, meaning that the pool will naturally bias towards it and the 32:1 "natural" ratio will skew in order to catch up on the over-represented task bucket.

In short, this preserves the local biasing and contention optimization of the batched queue while making the scheduler considerably better at self-tuning in the face of dynamic workloads.

File: core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala [hunk 1]
```diff
@@ -251,88 +248,75 @@ private final class WorkerThread(
...
+          if (element ne null) {
+            if (element.isInstanceOf[Array[IOFiber[_]]]) {
+              // The dequeued element was a batch of fibers. Enqueue the whole
+              // batch on the local queue and execute the first fiber.
+
+              // Make room for the batch if the local queue cannot accomodate
+              // all of the fibers as is.
+              queue.drainBatch(external, rnd)
+
+              val fiber = queue.enqueueBatch(element.asInstanceOf[Array[IOFiber[_]]])
```
Comment (djspiewak) [inline core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala]: Minor bit of silliness, but it's very likely that this will perform slightly better if you move the `asInstanceOf` up to the line immediately following the `isInstanceOf`. The reason for this is that HotSpot has a specific optimization to catch that pair of instructions and eliminate the check when doing the cast, which avoids double-checking and extra branching. By putting the instruction *here* you force the JVM to do a much more complex analysis that I'm not sure it's able to do. Would be worth testing.

**Review by djspiewak — COMMENTED** [review]
Looks very very good! I'm curious as to the performance results.

**Comment by vasilmkd** [comment]
Yeah, I was only looking to make this work. I'll do another pass with changes to comments and will address this too.

**Comment by vasilmkd** [comment]
Ok, this PR is ready in terms of changes. I would need to also re-do the local queue sampler mbean implementation, but I'm going to do that in another PR so that this one can remain somewhat focused. I'll still keep it as a draft, until we run some benchmarks.

File: core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala (hunk 1 above)
Comment (vasilmkd) [reply core/jvm/src/main/scala/cats/effect/unsafe/WorkerThread.scala]: Done. Did you know that `null.isInstanceOf[<any type here>]` always returns `false`?

**Comment by vasilmkd** [comment]
Running on my local machine:
`series/3.x`:
```
Benchmark                          (size)   Mode  Cnt   Score   Error    Units
WorkStealingBenchmark.scheduling  1000000  thrpt   20  12.259 ± 0.386  ops/min
```

`This PR`:
```
Benchmark                          (size)   Mode  Cnt   Score   Error    Units
WorkStealingBenchmark.scheduling  1000000  thrpt   20  12.026 ± 0.388  ops/min
```

**Comment by vasilmkd** [comment]
I'm running the benchmarks in a cloud instance, I will post the results later. I also plan to publish a snapshot and do the comparative benchmarks.

**Comment by vasilmkd** [comment]
16 core cloud instance scores (admittedly, this is not the most reliable source of data, but it's all I have).

`series/3.x`:
```
Benchmark                                             (cpuTokens)   (size)   Mode  Cnt      Score      Error    Units
AsyncBenchmark.async                                          N/A      100  thrpt   20  16822.231 ±  361.634    ops/s
AsyncBenchmark.bracket                                        N/A      100  thrpt   20  12522.736 ±  226.197    ops/s
AsyncBenchmark.cancelable                                     N/A      100  thrpt   20  15723.296 ±  226.794    ops/s
AsyncBenchmark.race                                           N/A      100  thrpt   20   4950.531 ±  153.045    ops/s
AsyncBenchmark.racePair                                       N/A      100  thrpt   20  20017.329 ± 1098.793    ops/s
AsyncBenchmark.start                                          N/A      100  thrpt   20   5371.408 ±  100.404    ops/s
AsyncBenchmark.uncancelable                                   N/A      100  thrpt   20  23653.999 ±  462.851    ops/s
AttemptBenchmark.errorRaised                                  N/A    10000  thrpt   20   1120.471 ±   39.738    ops/s
AttemptBenchmark.happyPath                                    N/A    10000  thrpt   20   1570.957 ±   51.658    ops/s
BothBenchmark.happyPath                                       N/A    10000  thrpt   20     21.422 ±    2.372    ops/s
DeepBindBenchmark.async                                       N/A    10000  thrpt   20   1132.839 ±   34.897    ops/s
DeepBindBenchmark.delay                                       N/A    10000  thrpt   20   3229.105 ±  107.623    ops/s
DeepBindBenchmark.pure                                        N/A    10000  thrpt   20   3706.581 ±   61.757    ops/s
DispatcherBenchmark.scheduling                                N/A     1000  thrpt   20     58.236 ±    2.022  ops/min
HandleErrorBenchmark.errorRaised                              N/A    10000  thrpt   20   1312.804 ±   20.846    ops/s
HandleErrorBenchmark.happyPath                                N/A    10000  thrpt   20   1402.251 ±   20.586    ops/s
MapCallsBenchmark.batch120                                    N/A      N/A  thrpt   20    282.756 ±    4.026    ops/s
MapCallsBenchmark.batch30                                     N/A      N/A  thrpt   20     73.749 ±    3.743    ops/s
MapCallsBenchmark.one                                         N/A      N/A  thrpt   20      2.767 ±    0.100    ops/s
MapStreamBenchmark.batch120                                   N/A      N/A  thrpt   20   1641.499 ±   38.237    ops/s
MapStreamBenchmark.batch30                                    N/A      N/A  thrpt   20    679.838 ±   11.031    ops/s
MapStreamBenchmark.one                                        N/A      N/A  thrpt   20   1052.260 ±   15.211    ops/s
ParallelBenchmark.parTraverse                               10000     1000  thrpt   20    259.628 ±    7.809    ops/s
ParallelBenchmark.traverse                                  10000     1000  thrpt   20     33.305 ±    0.031    ops/s
RaceBenchmark.happyPath                                       N/A    10000  thrpt   20     22.016 ±    1.430    ops/s
RedeemBenchmark.errorRaised                                   N/A    10000  thrpt   20    815.091 ±   31.252    ops/s
RedeemBenchmark.happyPath                                     N/A    10000  thrpt   20    952.050 ±   81.492    ops/s
RedeemWithBenchmark.errorRaised                               N/A    10000  thrpt   20    935.691 ±   13.046    ops/s
RedeemWithBenchmark.happyPath                                 N/A    10000  thrpt   20   1351.620 ±   43.377    ops/s
RefBenchmark.getAndUpdate                                     N/A      N/A  thrpt   20   1940.354 ±   69.377    ops/s
RefBenchmark.modify                                           N/A      N/A  thrpt   20   1745.435 ±   66.163    ops/s
ShallowBindBenchmark.async                                    N/A    10000  thrpt   20    864.954 ±   23.389    ops/s
ShallowBindBenchmark.delay                                    N/A    10000  thrpt   20   3196.206 ±   95.267    ops/s
ShallowBindBenchmark.pure                                     N/A    10000  thrpt   20   3644.993 ±  108.234    ops/s
WorkStealingBenchmark.alloc                                   N/A  1000000  thrpt   20      6.025 ±    0.178  ops/min
WorkStealingBenchmark.manyThreadsSchedulingBenchmark          N/A  1000000  thrpt   20     17.164 ±    0.325  ops/min
WorkStealingBenchmark.runnableScheduling                      N/A  1000000  thrpt   20    121.958 ±    1.545  ops/min
WorkStealingBenchmark.scheduling (*)                          N/A  1000000  thrpt   20      6.882 ±    0.787  ops/min
```

The `scheduling` score is suspiciously low, and I attribute it to the cloud instance performance. Check re-runs in later comments.

`This PR`:
```
Benchmark                                             (cpuTokens)   (size)   Mode  Cnt      Score      Error    Units
AsyncBenchmark.async                                          N/A      100  thrpt   20  16463.158 ±  249.595    ops/s
AsyncBenchmark.bracket                                        N/A      100  thrpt   20  12412.039 ±  143.258    ops/s
AsyncBenchmark.cancelable                                     N/A      100  thrpt   20  15881.182 ±  224.176    ops/s
AsyncBenchmark.race                                           N/A      100  thrpt   20   5282.131 ±  240.079    ops/s
AsyncBenchmark.racePair                                       N/A      100  thrpt   20  19844.984 ± 1053.497    ops/s
AsyncBenchmark.start                                          N/A      100  thrpt   20   4899.008 ±  407.314    ops/s
AsyncBenchmark.uncancelable                                   N/A      100  thrpt   20  23979.278 ±  554.653    ops/s
AttemptBenchmark.errorRaised                                  N/A    10000  thrpt   20   1082.108 ±   13.725    ops/s
AttemptBenchmark.happyPath                                    N/A    10000  thrpt   20   1553.057 ±   32.347    ops/s
BothBenchmark.happyPath                                       N/A    10000  thrpt   20     22.079 ±    1.422    ops/s
DeepBindBenchmark.async                                       N/A    10000  thrpt   20   1089.312 ±   44.340    ops/s
DeepBindBenchmark.delay                                       N/A    10000  thrpt   20   3240.221 ±   91.184    ops/s
DeepBindBenchmark.pure                                        N/A    10000  thrpt   20   3650.951 ±   59.833    ops/s
DispatcherBenchmark.scheduling                                N/A     1000  thrpt   20     58.608 ±    3.046  ops/min
HandleErrorBenchmark.errorRaised                              N/A    10000  thrpt   20   1310.049 ±   19.005    ops/s
HandleErrorBenchmark.happyPath                                N/A    10000  thrpt   20   1397.669 ±   19.320    ops/s
MapCallsBenchmark.batch120                                    N/A      N/A  thrpt   20    279.739 ±    5.304    ops/s
MapCallsBenchmark.batch30                                     N/A      N/A  thrpt   20     78.485 ±    2.171    ops/s
MapCallsBenchmark.one                                         N/A      N/A  thrpt   20      2.761 ±    0.114    ops/s
MapStreamBenchmark.batch120                                   N/A      N/A  thrpt   20   1640.282 ±   45.051    ops/s
MapStreamBenchmark.batch30                                    N/A      N/A  thrpt   20    681.799 ±    7.647    ops/s
MapStreamBenchmark.one                                        N/A      N/A  thrpt   20   1079.243 ±   21.200    ops/s
ParallelBenchmark.parTraverse                               10000     1000  thrpt   20    271.972 ±    7.064    ops/s
ParallelBenchmark.traverse                                  10000     1000  thrpt   20     33.353 ±    0.044    ops/s
RaceBenchmark.happyPath                                       N/A    10000  thrpt   20     22.783 ±    1.568    ops/s
RedeemBenchmark.errorRaised                                   N/A    10000  thrpt   20    773.892 ±   18.937    ops/s
RedeemBenchmark.happyPath                                     N/A    10000  thrpt   20    981.409 ±   74.080    ops/s
RedeemWithBenchmark.errorRaised                               N/A    10000  thrpt   20   1012.925 ±   18.995    ops/s
RedeemWithBenchmark.happyPath                                 N/A    10000  thrpt   20   1359.581 ±   37.402    ops/s
RefBenchmark.getAndUpdate                                     N/A      N/A  thrpt   20   1939.056 ±   70.017    ops/s
RefBenchmark.modify                                           N/A      N/A  thrpt   20   1803.804 ±   49.859    ops/s
ShallowBindBenchmark.async                                    N/A    10000  thrpt   20    845.795 ±    7.458    ops/s
ShallowBindBenchmark.delay                                    N/A    10000  thrpt   20   3286.053 ±   92.127    ops/s
ShallowBindBenchmark.pure                                     N/A    10000  thrpt   20   3773.739 ±  157.648    ops/s
WorkStealingBenchmark.alloc                                   N/A  1000000  thrpt   20      6.412 ±    0.132  ops/min
WorkStealingBenchmark.manyThreadsSchedulingBenchmark          N/A  1000000  thrpt   20     13.479 ±    0.277  ops/min
WorkStealingBenchmark.runnableScheduling                      N/A  1000000  thrpt   20    112.979 ±    1.342  ops/min
WorkStealingBenchmark.scheduling                              N/A  1000000  thrpt   20      9.307 ±    1.710  ops/min
```

I ran the `WorkStealingBenchmark` again on this PR:
```
Benchmark                                              (size)   Mode  Cnt    Score   Error    Units
WorkStealingBenchmark.alloc                           1000000  thrpt   20    6.306 ± 0.182  ops/min
WorkStealingBenchmark.manyThreadsSchedulingBenchmark  1000000  thrpt   20   13.094 ± 0.297  ops/min
WorkStealingBenchmark.runnableScheduling              1000000  thrpt   20  114.017 ± 4.936  ops/min
WorkStealingBenchmark.scheduling                      1000000  thrpt   20   11.420 ± 2.104  ops/min
```

**Comment by vasilmkd** [comment]
I released a snapshot version:
```scala
"io.vasilev" %%% "cats-effect" % "3.3-151-7de2ca2"
```

**Comment by vasilmkd** [comment]
Comparative benchmarks run locally:
`3.2.8`:
```
Benchmark                                  Mode  Cnt      Score    Error    Units
Benchmarks.catsEffect3Alloc               thrpt   20      3.753 ±  0.022  ops/min
Benchmarks.catsEffect3DeepBind            thrpt   20  16738.508 ± 43.263    ops/s
Benchmarks.catsEffect3EnqueueDequeue      thrpt   20    465.600 ±  2.678    ops/s
Benchmarks.catsEffect3RuntimeChainedFork  thrpt   20   2582.992 ± 33.598    ops/s
Benchmarks.catsEffect3RuntimeForkMany     thrpt   20    281.838 ±  1.853    ops/s
Benchmarks.catsEffect3RuntimePingPong     thrpt   20   1325.878 ± 19.645    ops/s
Benchmarks.catsEffect3RuntimeYieldMany    thrpt   20    201.913 ± 11.977    ops/s
Benchmarks.catsEffect3Scheduling          thrpt   20     12.305 ±  0.442  ops/min
```

`snapshot from this PR`:
```
Benchmark                                  Mode  Cnt      Score     Error    Units
Benchmarks.catsEffect3Alloc               thrpt   20      3.729 ±   0.127  ops/min
Benchmarks.catsEffect3DeepBind            thrpt   20  16564.685 ± 101.847    ops/s
Benchmarks.catsEffect3EnqueueDequeue      thrpt   20    460.296 ±   2.832    ops/s
Benchmarks.catsEffect3RuntimeChainedFork  thrpt   20   2680.373 ±  14.815    ops/s
Benchmarks.catsEffect3RuntimeForkMany     thrpt   20    282.285 ±   5.081    ops/s
Benchmarks.catsEffect3RuntimePingPong     thrpt   20   1299.313 ±  20.316    ops/s
Benchmarks.catsEffect3RuntimeYieldMany    thrpt   20    217.685 ±  17.916    ops/s
Benchmarks.catsEffect3Scheduling          thrpt   20     12.499 ±   0.466  ops/min
```

**Comment by vasilmkd** [comment]
Comparative benchmarks on the 16 core cloud instance:
`3.2.8`:
```
Benchmark                                  Mode  Cnt     Score    Error    Units
Benchmarks.catsEffect3Alloc               thrpt   20     5.617 ±  0.151  ops/min
Benchmarks.catsEffect3DeepBind            thrpt   20  5215.017 ± 49.818    ops/s
Benchmarks.catsEffect3EnqueueDequeue      thrpt   20   234.701 ±  8.005    ops/s
Benchmarks.catsEffect3RuntimeChainedFork  thrpt   20  1368.123 ± 49.539    ops/s
Benchmarks.catsEffect3RuntimeForkMany     thrpt   20   130.754 ±  4.686    ops/s
Benchmarks.catsEffect3RuntimePingPong     thrpt   20   531.563 ± 28.601    ops/s
Benchmarks.catsEffect3RuntimeYieldMany    thrpt   20    81.483 ±  7.074    ops/s
Benchmarks.catsEffect3Scheduling          thrpt   20     9.924 ±  0.493  ops/min
```

`snapshot from this PR`:
```
Benchmark                                  Mode  Cnt     Score    Error    Units
Benchmarks.catsEffect3Alloc               thrpt   20     6.341 ±  0.162  ops/min
Benchmarks.catsEffect3DeepBind            thrpt   20  5243.657 ± 74.826    ops/s
Benchmarks.catsEffect3EnqueueDequeue      thrpt   20   241.071 ±  1.306    ops/s
Benchmarks.catsEffect3RuntimeChainedFork  thrpt   20  1342.761 ± 17.865    ops/s
Benchmarks.catsEffect3RuntimeForkMany     thrpt   20   137.492 ±  5.717    ops/s
Benchmarks.catsEffect3RuntimePingPong     thrpt   20   514.737 ± 33.954    ops/s
Benchmarks.catsEffect3RuntimeYieldMany    thrpt   20    84.721 ± 15.061    ops/s
Benchmarks.catsEffect3Scheduling          thrpt   20    11.608 ±  0.779  ops/min
```

**Comment by djspiewak** [comment]
This looks outstanding! Really it does. We don't really have a benchmark which illustrates dynamic fairness tuning (I'm thinking about it), so that's not a case we measure particularly well right now. The fact that we were able to make this change and the existing throughput measurements are all within the margin of error says to me that this is just an unambiguous win.

**Comment by vasilmkd** [comment]
@djspiewak I don't plan on making other changes to this PR. Can I mark it as ready for review?

**Review by djspiewak — APPROVED** [review]
